modules/core/src/main/java/org/locationtech/jts/geom/GeometryCollectionDimension.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (c) 2024 Martin Davis.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * and Eclipse Distribution License v. 1.0 which accompanies this distribution.
+ * The Eclipse Public License is available at http://www.eclipse.org/legal/epl-v20.html
+ * and the Eclipse Distribution License is available at
+ *
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ */
+package org.locationtech.jts.geom;
+
+import java.util.Iterator;
+
+/**
+ * Computes and caches dimension information for {@link GeometryCollection}s.
+ * Optimizes performance of dimension reporting for heterogeneous collections.
+ * 
+ * @author mdavis
+ */
+class GeometryCollectionDimension {
+  private int dimension = Dimension.FALSE;
+  private boolean hasP = false;
+  private boolean hasL = false;
+  private boolean hasA = false;
+
+  public GeometryCollectionDimension(GeometryCollection coll) {
+    init(coll);
+  }
+
+  private void init(GeometryCollection coll) {
+    Iterator geomi = new GeometryCollectionIterator(coll);
+    while (geomi.hasNext()) {
+      Geometry elem = (Geometry) geomi.next();
+      //-- empty elements still determine dimension, to match previous semantics
+      if (elem instanceof Point) {
+        hasP = true;
+        if (dimension < Dimension.P) dimension = Dimension.P;
+      }
+      if (elem instanceof LineString) {
+        hasL = true;
+        if (dimension < Dimension.L) dimension = Dimension.L;
+      }
+      if (elem instanceof Polygon) {
+        hasA = true;
+        if (dimension < Dimension.A) dimension = Dimension.A;
+      }
+    }
+  }
+  
+  public boolean hasDimension(int dim) {
+    if (dim == Dimension.A && hasA) return true;
```

Comment:
I don't understand the `dim == x` check here. If we have a mixed polygon/line collection, isn't `dim` equal to `Dimension.A` but `hasDimension(Dimension.L)` should be true?

Reply:
`dim` is the argument to `hasDimension`.  So for `hasDimension(Dimension.L)` `dim` is 1, and that matches 
```
dim == Dimension.L && hasL
```

Actually this logic would be simpler as:
```
    if (dim == Dimension.A) return hasA;
    if (dim == Dimension.L) return hasL;
    if (dim == Dimension.P) return hasP;
    return false;
```

Reply:
Or even better:
```
switch (dim) {
case Dimension.A: return hasA;
case Dimension.L: return hasL;
case Dimension.P: return hasP;
}
return false;
```

Reply:
Thank you, I was pretty sure I was having a brain failure of sorts.

Reply:
See https://github.com/locationtech/jts/commit/9133b4717ce29cb20c2f21894a645ff1a44cd74f